### PR TITLE
[ELY-788] SASL mechs null handling

### PR DESCRIPTION
**Pull request by honza889**
https://issues.jboss.org/browse/ELY-788
(no subsystem part)

### CODE REVIEW

File: src/main/java/org/wildfly/security/sasl/digest/DigestClientFactory.java [hunk 1]
```diff
@@ -60,6 +62,9 @@ public DigestClientFactory(final Provider provider) {
     public SaslClient createSaslClient(String[] mechanisms, String authorizationId, String protocol, String serverName,
             Map<String, ?> props, CallbackHandler cbh) throws SaslException {
 
+        if (cbh == null) throw ElytronMessages.log.mechCallbackHandlerMissing().toSaslException();
```
Comment (dmlloyd) [inline src/main/java/org/wildfly/security/sasl/digest/DigestClientFactory.java]: Null parameter checks should always use "Assert.checkNotNullParam" if possible.

Comment (honza889) [inline src/main/java/org/wildfly/security/sasl/digest/DigestClientFactory.java]: Yes, but @kwart requested Sasl should throw SaslException only.

Comment (dmlloyd) [inline src/main/java/org/wildfly/security/sasl/digest/DigestClientFactory.java]: I don't think this is a realistic requirement; RuntimeExceptions are unavoidable and are preferred in the event of a programming error condition (like a null parameter).  This allows programming errors to take an exceptional path on the client or server.

Comment (kwart) [inline src/main/java/org/wildfly/security/sasl/digest/DigestClientFactory.java]: We should keep the compatibility with the JDK implementation. Some mechanisms (e.g. `DIGEST-MD5`) reimplement the ones provided by the JDK and the exception handling  should be kept as close as possible to this default.

Example from [com.sun.security.sasl.digest.FactoryImpl class](http://hg.openjdk.java.net/jdk8u/jdk8u/jdk/file/0e4fc29a5ce4/src/share/classes/com/sun/security/sasl/digest/FactoryImpl.java):

```java
if (cbh == null) {
    throw new SaslException(
        "Callback handler with support for RealmChoiceCallback, " +
        "RealmCallback, NameCallback, and PasswordCallback " +
        "required");
}
```

Comment (darranl) [inline src/main/java/org/wildfly/security/sasl/digest/DigestClientFactory.java]: I am not sure the compatibility is as important at this point - in both cases the end result is if you don't provide the required parameters the mechanism is unusable.

Comment (dmlloyd) [inline src/main/java/org/wildfly/security/sasl/digest/DigestClientFactory.java]: Some of the JDK implementations also throw NPE if a parameter is null; essentially it's pretty arbitrary and not a part of the API contract defined by either Oracle, and we define our method call contract to throw IAE with a standard i18n exception message and code when a parameter is null due to a programming error.  This makes debugging easier and also helps manage the user's expectations.

**Comment by honza889** [comment]
Changed to use Assert.checkNotNullParam - fixed by @dmlloyd notes.

**Pull request merged**